src/mlpack/methods/rbm/rbm_impl.hpp

```diff
@@ -0,0 +1,206 @@
+  if (!useMonitoringCost)
+  {
+    // Do not use unsafe col predictor as we change the input
+    Gibbs(std::move(predictors.col(i)), std::move(negativeSamples));
+    return (FreeEnergy(std::move(predictors.col(i))) -
+        FreeEnergy(std::move(negativeSamples)));
+  }
+  else
+  {
+    if (persistence)
+    {
+      size_t idx = RandInt(0, predictors.n_rows);
+      corruptInput.row(idx) = 1 - corruptInput.row(idx);
+      return std::log(LogisticFunction::Fn(FreeEnergy(std::move(corruptInput)) -
+          FreeEnergy(std::move(arma::round(predictors.col(i)))))) *
+          predictors.n_rows;
+    }
+    else
+    {
+      // Cross entropy loss
+      rbmPolicy.VisibleLayer().Forward(std::move(predictors.col(i)),
+        std::move(preActivation));
+      preActivation = -preActivation;
+      SoftplusFunction::Fn(preActivation, softOutput);
+      softOutput = -softOutput;
+      preActivation = -preActivation - 1;
+      SoftplusFunction::Fn(preActivation, softOutput);
+      softOutput = -softOutput;
+      return arma::accu(
+          predictors.col(i) * softOutput.t() + (1 - predictors.col(i)) *
+          preActivation.t());
+    }
+  }
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::
+    SampleHidden(arma::mat&& input, arma::mat&& output)
+{
+  rbmPolicy.SampleHidden(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::
+    SampleVisible(arma::mat&& input, arma::mat&& output)
+{
+  rbmPolicy.SampleVisible(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::
+    Gibbs(arma::mat&& input, arma::mat&& output, size_t steps)
+{
+  steps = (steps == SIZE_MAX) ? this-> numSteps: steps;
+
+  if (persistence && !state.is_empty())
+  {
+    SampleHidden(std::move(state), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  else
+  {
+    SampleHidden(std::move(input), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+
+  for (size_t j = 1; j < steps; j++)
+  {
+    SampleHidden(std::move(output), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  if (persistence)
+    state = output;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::
+    Gradient(arma::mat& /*parameters*/, const size_t input, arma::mat& output)
+{
+  // Collect the negative samples
+  rbmPolicy.PositivePhase(std::move(predictors.col(input)));
+  for (size_t i = 0; i < mSteps; i++)
+  {
+    Gibbs(std::move(predictors.col(input)), std::move(tempNegativeSmaples));
+    negativeSamples.col(i) = tempNegativeSmaples;
+  }
+  rbmPolicy.NegativePhase(std::move(negativeSamples));
```

Comment:
if i call NegativePhase() mSteps number of times. That would mean that i would be storing the negativeGradient in a tempvariable(since negativeGradient is overwritten) and then divide it by mSteps. How is this better than the previous approach.